build/wpt_test.bzl

```diff
@@ -0,0 +1,110 @@
+load("//:build/wd_test.bzl", "wd_test")
+
+# The public entry point is a macro named wpt_test. It first invokes a private
+# rule named _wpt_test_gen to access the files in the wpt filegroup and
+# generate a corresponding wd-test file. It then invokes the wd_test macro
+# to set up the test.
+
+def wpt_test(name, wpt_directory, test_js):
+    test_gen_rule = "{}@_wpt_test_gen".format(name)
+    _wpt_test_gen(
+        name = test_gen_rule,
+        test_name = name,
+        wpt_directory = wpt_directory,
+        test_js = test_js,
+    )
+
+    wd_test(
+        name = "{}".format(name),
+        src = test_gen_rule,
+        args = ["--experimental"],
+        data = [
+            "//src/wpt:wpt-test-harness",
+            test_js,
+            wpt_directory,
+            "//src/workerd/io:trimmed-supported-compatibility-date.txt",
+        ],
+    )
+
+def _wpt_test_gen_impl(ctx):
+    src = ctx.actions.declare_file("{}.wd-test".format(ctx.attr.test_name))
+    ctx.actions.write(
+        output = src,
+        content = WPT_TEST_TEMPLATE.format(
+            test_name = ctx.attr.test_name,
+            test_js = wd_relative_path(ctx.file.test_js),
+            modules = generate_external_modules(ctx.attr.wpt_directory.files),
+        ),
+    )
+
+    return DefaultInfo(
+        files = depset([src]),
+    )
+
+WPT_TEST_TEMPLATE = """
+using Workerd = import "/workerd/workerd.capnp";
+const unitTests :Workerd.Config = (
+  services = [
+    ( name = "{test_name}",
+      worker = (
+        modules = [
+          (name = "worker", esModule = embed "{test_js}"),
+          (name = "harness", esModule = embed "../../../../../workerd/src/wpt/harness.js"),
+          {modules}
+        ],
+        bindings = [
+          (name = "wpt", service = "wpt"),
+        ],
+        compatibilityDate = embed "../../../../../workerd/src/workerd/io/trimmed-supported-compatibility-date.txt",
+        compatibilityFlags = ["nodejs_compat_v2", "experimental"],
```

Comment:
Nit: if the compat date is after 2024-09-23 then you can just use `nodejs_compat` instead of `nodejs_compat_v2`